lfs/transfer_queue.go

```diff
@@ -0,0 +1,212 @@
+	Check() (*objectResource, *WrappedError)
+	Transfer(CopyCallback) *WrappedError
+	Object() *objectResource
+	Oid() string
+	Size() int64
+	SetObject(*objectResource)
+}
+
+// TransferQueue provides a queue that will allow concurrent transfers.
+type TransferQueue struct {
+	transferc        chan Transferable
+	errorc           chan *WrappedError
+	errors           []*WrappedError
+	wg               sync.WaitGroup
+	workers          int
+	files            int
+	finished         int64
+	size             int64
+	authCond         *sync.Cond
+	transferables    map[string]Transferable
+	bar              *pb.ProgressBar
+	clientAuthorized int32
+	transferKind     string
+}
+
+// newTransferQueue builds a TransferQueue, allowing `workers` concurrent transfers.
+func newTransferQueue(workers, files int) *TransferQueue {
+	return &TransferQueue{
+		transferc:     make(chan Transferable, files),
+		errorc:        make(chan *WrappedError),
+		workers:       workers,
+		files:         files,
+		authCond:      sync.NewCond(&sync.Mutex{}),
+		transferables: make(map[string]Transferable),
+	}
+}
+
+// Add adds a Transferable to the transfer queue.
+func (q *TransferQueue) Add(t Transferable) {
+	q.transferables[t.Oid()] = t
+}
+
+// processIndividual processes the queue of transfers one at a time by making
+// a POST call for each object, feeding the results to the transfer workers.
+// If configured, the object transfers can still happen concurrently, the
+// sequential nature here is only for the meta POST calls.
+func (q *TransferQueue) processIndividual() {
+	apic := make(chan Transferable, q.files)
+	workersReady := make(chan int, q.workers)
+	var wg sync.WaitGroup
+
+	for i := 0; i < q.workers; i++ {
+		go func() {
+			workersReady <- 1
+			for t := range apic {
+				// If an API authorization has not occured, we wait until we're woken up.
+				q.authCond.L.Lock()
+				if atomic.LoadInt32(&q.clientAuthorized) == 0 {
+					q.authCond.Wait()
+				}
+				q.authCond.L.Unlock()
+
+				obj, err := t.Check()
+				if err != nil {
+					q.errorc <- err
+					wg.Done()
+					continue
+				}
+				if obj != nil {
+					q.wg.Add(1)
+					t.SetObject(obj)
+					q.transferc <- t
+				}
+				wg.Done()
+			}
+		}()
+	}
+
+	q.bar.Prefix(fmt.Sprintf("(%d of %d files) ", q.finished, len(q.transferables)))
+	q.bar.Start()
+
+	for _, t := range q.transferables {
+		wg.Add(1)
+		apic <- t
+	}
+
+	<-workersReady
+	q.authCond.Signal() // Signal the first goroutine to run
+	close(apic)
+	wg.Wait()
+
+	close(q.transferc)
+}
+
+// processBatch processes the queue of transfers using the batch endpoint,
+// making only one POST call for all objects. The results are then handed
+// off to the transfer workers.
+func (q *TransferQueue) processBatch() {
+	q.files = 0
+	transfers := make([]*objectResource, 0, len(q.transferables))
+	for _, t := range q.transferables {
+		transfers = append(transfers, &objectResource{Oid: t.Oid(), Size: t.Size()})
+	}
+
+	objects, err := Batch(transfers)
```

Comment:
We should define an upper bound for the number of transfers to batch at a time. This way the client isn't sending massive json payloads that servers have to process. We don't want to burden servers with the prospect of handling POST payloads with thousands of objects.